### PR TITLE
[mono] Enable debugging for netcore configurations by default.

### LINKED ISSUE #31662: Enable --debug mode in Mono by default
Historically Mono required passing the `--debug` flag to enable reading of .pdb files for getting line numbers in stack traces.

Since .NET Core does this by default we should look at enabling this on Mono too.

/cc @lambdageek @steveisok @marek-safar

### PR DESCRIPTION
**Pull request by steveisok**
What this means is:

1. --debug is no longer needed since the default is TRUE
2. --debug=ignore is a new switch to go back to the old mono default

Fixes https://github.com/dotnet/runtime/issues/31662

### CODE REVIEW

File: src/mono/mono/mini/driver.c [hunk 1]
```diff
@@ -1605,7 +1610,11 @@ mini_usage (void)
 		"\n"
 		"Development:\n"
 		"    --aot[=<options>]      Compiles the assembly to native code\n"
+#ifndef ENABLE_NETCORE
+		"    --debug=ignore    Disable debugging support, use --help-debug for details\n"
+#else
 		"    --debug[=<options>]    Enable debugging support, use --help-debug for details\n"
+#endif
```
Comment (vargaz) [inline src/mono/mono/mini/driver.c]: Note that the old --debug had all kinds of options to enable which have a perf cost like --debug=casts.

File: src/mono/mono/mini/mini-runtime.c [hunk 2]
```diff
@@ -3662,6 +3662,10 @@ mini_parse_debug_option (const char *option)
 
 	if (!strcmp (option, "handle-sigint"))
 		mini_debug_options.handle_sigint = TRUE;
+#ifdef ENABLE_NETCORE
+	else if (!strcmp (option, "ignore"))
+		mini_debug_options.enabled = FALSE;
+#endif
```
Comment (vargaz) [inline src/mono/mono/mini/mini-runtime.c]: This handles the MONO_DEBUG env variable which contains runtime debugging options.

File: src/mono/mono/mini/driver.c (hunk 1 above)
Comment (steveisok) [reply src/mono/mono/mini/driver.c]: It's still there - I messed up the printout.

File: src/mono/mono/mini/driver.c [hunk 3]
```diff
@@ -1605,7 +1610,12 @@ mini_usage (void)
 		"\n"
 		"Development:\n"
 		"    --aot[=<options>]      Compiles the assembly to native code\n"
+#ifdef ENABLE_NETCORE
```
Comment (jaykrell) [inline src/mono/mono/mini/driver.c]: I suggest -nodebug.

File: src/mono/mono/mini/driver.c [hunk 4]
```diff
@@ -2325,12 +2339,21 @@ mono_main (int argc, char* argv[])
 			mname = argv [++i];
 			mono_graph_options = MONO_GRAPH_CFG;
 			action = DO_DRAW;
+#ifndef ENABLE_NETCORE
 		} else if (strcmp (argv [i], "--debug") == 0) {
 			enable_debugging = TRUE;
+#endif
```
Comment (akoeplinger) [inline src/mono/mono/mini/driver.c]: I think we should still support passing `--debug` in netcore mode even though it is a no-op, just for compatibilities sake (e.g. having a MONO_ENV_OPTIONS set)

File: src/mono/mono/mini/driver.c (hunk 1 above)
Comment (akoeplinger) [reply src/mono/mono/mini/driver.c]: We should add the new option here too: https://github.com/dotnet/runtime/blob/a7f8cdc0de945561ae3fcaa72440c0ec56d69060/src/mono/mono/mini/driver.c#L1666-L1668

**Review by akoeplinger — CHANGES_REQUESTED** [review]

File: src/mono/mono/mini/driver.c (hunk 4 above)
Comment (jaykrell) [reply src/mono/mono/mini/driver.c]: I still like `-nodebug` or maybe `-optimize=nodebug` but so verbose.

**Comment by marek-safar** [comment]
Is this really the approach we want to take? I don't see an easy way for every release build to disable this behaviour.

We could instead recognize existing `COMPlus_EnableDiagnostics` and convert it to full MonoVM debug mode

File: src/mono/mono/mini/driver.c [hunk 5]
```diff
@@ -216,6 +216,11 @@ parse_debug_options (const char* p)
 		} else if (!strncmp (p, "gdb", 3)) {
 			opt->gdb = TRUE;
 			p += 3;
+#ifdef ENABLE_NETCORE
+		} else if (!strncmp (p, "ignore", 6)) {
+			opt->enabled = FALSE;
```
Comment (lewurm) [inline src/mono/mono/mini/driver.c]: what is setting `opt->enabled = TRUE`?

**Comment by lewurm** [comment]
> Is this really the approach we want to take? I don't see an easy way for every release build to disable this behaviour.

For Release the tooling can pass 
1. `--debug=ignore`, or
2. Don't care, because the performance implications aren't that bad. Having useful stack traces is imho the better default behavior anyway.

**Comment by akoeplinger** [comment]
Additionally both XI and VSMac already unconditionally pass `--debug`:

- https://github.com/xamarin/xamarin-macios/blob/5c4b339ed320c474f8a88a3be4435274636ad38d/tools/mtouch/mtouch.cs#L390
- https://github.com/xamarin/vsmac/blob/68e5a9cb133ee5aff44697a3c678e5ad82f0548e/main/build/MacOSX/monostub.mm#L338

**Comment by steveisok** [comment]
@marek-safar What do we want to do?  I can revert / do what you suggested.  However, if there are no/limited performance implications, then my vote is to keep it as is.

**Review by marek-safar — APPROVED** [review]
@vargaz confirmed the simple version of `--debug` is not that expensive so let's enable it by default.

File: src/mono/mono/mini/driver.c (hunk 5 above)
Comment (steveisok) [reply src/mono/mono/mini/driver.c]: I set it to true at the start of the function

**Review by lewurm — APPROVED** [review]

**Review by akoeplinger — APPROVED** [review]

**Pull request merged**